velox/functions/lib/window/NthValue.cpp

```diff
@@ -28,39 +28,46 @@ class NthValueFunction : public exec::WindowFunction {
       const std::vector<exec::WindowFunctionArg>& args,
       const TypePtr& resultType,
       bool ignoreNulls,
+      const TypePtr& offsetType,
       velox::memory::MemoryPool* pool)
       : WindowFunction(resultType, pool, nullptr), ignoreNulls_(ignoreNulls) {
     VELOX_CHECK_EQ(args.size(), 2);
     VELOX_CHECK_NULL(args[0].constantValue);
     valueIndex_ = args[0].index.value();
-    if (args[1].type->isInteger()) {
-      VELOX_USER_CHECK(
-          args[1].constantValue, "Offset must be literal for spark");
+    if (args[1].constantValue) {
       if (args[1].constantValue->isNullAt(0)) {
         isConstantOffsetNull_ = true;
         return;
       }
-      constantOffset_ =
-          args[1]
-              .constantValue->template as<ConstantVector<int32_t>>()
-              ->valueAt(0);
-      VELOX_USER_CHECK_GE(
-          constantOffset_.value(), 1, "Offset must be at least 1");
-    } else {
-      if (args[1].constantValue) {
-        if (args[1].constantValue->isNullAt(0)) {
-          isConstantOffsetNull_ = true;
-          return;
-        }
+      if (offsetType->isInteger()) {
+        constantOffset_ =
+            args[1]
+                .constantValue->template as<ConstantVector<int32_t>>()
+                ->valueAt(0);
+        VELOX_USER_CHECK_GE(
+            std::get<int32_t>(constantOffset_.value()),
+            1,
+            "Offset must be at least 1");
+      } else if (offsetType->isBigint()) {
         constantOffset_ =
             args[1]
                 .constantValue->template as<ConstantVector<int64_t>>()
                 ->valueAt(0);
         VELOX_USER_CHECK_GE(
-            constantOffset_.value(), 1, "Offset must be at least 1");
+            std::get<int64_t>(constantOffset_.value()),
+            1,
+            "Offset must be at least 1");
       } else {
-        offsetIndex_ = args[1].index.value();
-        offsets_ = BaseVector::create<FlatVector<int64_t>>(BIGINT(), 0, pool);
+        VELOX_FAIL("Invalid offset type: {}", offsetType->toString());
+      }
+    } else {
+      offsetIndex_ = args[1].index.value();
+      if (offsetType->isInteger()) {
+        offsets_ = BaseVector::create<FlatVector<int32_t>>(offsetType, 0, pool);
+      } else if (offsetType->isBigint()) {
+        offsets_ = BaseVector::create<FlatVector<int64_t>>(offsetType, 0, pool);
+      } else {
+        VELOX_FAIL("Invalid offset type: {}", offsetType->toString());
```

Comment:
L61 and L70 checks can be moved to the begining of the function VELOX_CHECK(type is int || type bigint)